src/collapse/ghcprof.rs

```diff
@@ -0,0 +1,240 @@
+
+use log::warn;
+
+use crate::collapse::common::Occurrences;
+use crate::collapse::Collapse;
+
+// These are the identifying words of the callgraph table, note that ticks and bytes columns are optional so not present
+static START_LINE: &[&str] = &[
+    "COST", "CENTRE", "MODULE", "SRC", "no.", "entries", "%time", "%alloc", "%time", "%alloc",
+];
+
+/// `ghcprof` folder configuration options.
+#[derive(Clone, Debug, Default)]
+#[non_exhaustive]
+pub struct Options {
+    /// Column to source associated value from, default is `Source::PercentTime`.
+    pub source: Source,
+}
+
+/// Which prof column to use as the cost centre of the output stacks
+#[derive(Clone, Debug, Default)]
+#[non_exhaustive]
+pub enum Source {
+    #[default]
+    /// The indivial %time column representing individual time as a percent of the total
+    PercentTime,
+    /// The ticks column representing individual runtime ticks
+    Ticks,
+    /// The bytes column representing individual bytes allocated
+    Bytes,
+}
+
+/// A stack collapser for the output of `ghc`'s prof files.
+///
+/// To construct one, either use `ghcprof::Folder::default()` or create an [`Options`] and use
+/// `ghcprof::Folder::from(options)`.
+#[derive(Clone, Default)]
+pub struct Folder {
+    /// Cost for the current stack frame.
+    current_cost: usize,
+
+    /// Function on the stack in this entry thus far.
+    stack: Vec<String>,
+
+    opt: Options,
+}
+
+// The starting character offset of important columns
+#[derive(Debug)]
+struct Cols {
+    cost_centre: usize,
+    module: usize,
+    source: usize,
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: io::BufRead,
+        W: io::Write,
+    {
+        // Consume the header...
+        let mut line = Vec::new();
+        let cols = loop {
+            line.clear();
+            if reader.read_until(b'\n', &mut line)? == 0 {
+                warn!("File ended before start of call graph");
+                return Ok(());
+            };
+            let l = String::from_utf8_lossy(&line);
+
+            if l.split_whitespace()
+                .take(START_LINE.len())
+                .eq(START_LINE.iter().cloned())
+            {
+                let cost_centre = 0;
+                let module = l.find("MODULE").unwrap_or(0);
+                // Pick out these fixed columns, first two are individual only
+                // "%time %alloc   %time %alloc"
+                // `ticks` and `bytes` columns are optional and might appear on the end
+                // ticks header is right aligned
+                // bytes header is right aligned
+                //   - BUT it has a max width of 9 whilst its values can exceed (but are always space separted)
+                // "%time %alloc   %time %alloc  ticks  bytes"
+                let source = match self.opt.source {
+                    Source::PercentTime => l.find("%time").unwrap(),
```

Comment:
This still panics, but should return an error instead (like the other two now do).

Reply:
it's one of those cases where we "know" it's there since otherwise we couldn't be in the block because we just matched it as part of `START_LINE`. I've added a nice return error anyway and made clippy happy

Reply:
Ah, in that case you can use `expect` instead and give the rationale!

Reply:
okay swapped it